Resign Now Playing status when WKWebView suspends all media playback

jernoble opened the pull request:
#### 07fd7c7499a7c38e9794fe517374f03777bbe3db
<pre>
Resign Now Playing status when WKWebView suspends all media playback
<a href="https://bugs.webkit.org/show_bug.cgi?id=276133">https://bugs.webkit.org/show_bug.cgi?id=276133</a>
<a href="https://rdar.apple.com/129853095">rdar://129853095</a>

Reviewed by Eric Carlson.

When a WKWebView suspensd all media playback, we should resign Now Playing status (because
we can no longer resume playback from Now Playing controls). Additionally, when the page is
ephemeral, we should not report the media metadata information to Now Playing.

* LayoutTests/media/now-playing-info-media-session-private-browsing-expected.txt: Added.
* LayoutTests/media/now-playing-info-media-session-private-browsing.html: Added.
* LayoutTests/media/now-playing-info-media-session-suspend-playback-expected.txt: Added.
* LayoutTests/media/now-playing-info-media-session-suspend-playback.html: Added.
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::isNowPlayingEligible const):
(WebCore::AudioContext::nowPlayingInfo const):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::isNowPlayingEligible const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::computeNowPlayingInfo const):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateNowPlayingInfo):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::suspendAllMediaPlayback):
(WebCore::Internals::resumeAllMediaPlayback):
(WebCore::Internals::nowPlayingMetadata const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/280914@main">https://commits.webkit.org/280914@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d4948ce588e0ff49f9daeee911c4321314df7c60

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58004 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37332 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/10481 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/61629 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8449 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60132 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/44968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/8638 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/47009 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6019 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60034 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35011 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50138 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/27840 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/31775 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7433 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7453 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/53722 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/7701 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63317 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/1918 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/7772 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/54233 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/1925 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50149 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54373 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12832 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/1644 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33161 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34247 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35331 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/33992 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

eric-carlson commented on Source/WebCore/html/MediaElementSession.cpp:
```diff
@@ -1271,6 +1271,8 @@ std::optional<NowPlayingInfo> MediaElementSession::computeNowPlayingInfo() const
+    if (!page)
+        return { };
 
     bool allowsNowPlayingControlsVisibility = page && !page->isVisibleAndActive();
```
Nit: you can remove the `page &&`

jernoble committed:
Resign Now Playing status when WKWebView suspends all media playback
https://bugs.webkit.org/show_bug.cgi?id=276133
rdar://129853095

Reviewed by Eric Carlson.

When a WKWebView suspensd all media playback, we should resign Now Playing status (because
we can no longer resume playback from Now Playing controls). Additionally, when the page is
ephemeral, we should not report the media metadata information to Now Playing.

* LayoutTests/media/now-playing-info-media-session-private-browsing-expected.txt: Added.
* LayoutTests/media/now-playing-info-media-session-private-browsing.html: Added.
* LayoutTests/media/now-playing-info-media-session-suspend-playback-expected.txt: Added.
* LayoutTests/media/now-playing-info-media-session-suspend-playback.html: Added.
* Source/WebCore/Modules/webaudio/AudioContext.cpp:
(WebCore::AudioContext::isNowPlayingEligible const):
(WebCore::AudioContext::nowPlayingInfo const):
* Source/WebCore/html/HTMLMediaElement.cpp:
(WebCore::HTMLMediaElement::isNowPlayingEligible const):
* Source/WebCore/html/HTMLMediaElement.h:
* Source/WebCore/html/MediaElementSession.cpp:
(WebCore::MediaElementSession::computeNowPlayingInfo const):
* Source/WebCore/platform/audio/cocoa/MediaSessionManagerCocoa.mm:
(WebCore::MediaSessionManagerCocoa::updateNowPlayingInfo):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::suspendAllMediaPlayback):
(WebCore::Internals::resumeAllMediaPlayback):
(WebCore::Internals::nowPlayingMetadata const):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/280914@main
